include/RAJA/policy/hip/atomic.hpp

```diff
@@ -55,200 +55,244 @@ using hip_atomicCommon_builtin_types = ::camp::list<
+#else
+                       unsigned long long>>>;
+#endif
+};
 
-template <typename T,
-          std::enable_if_t<!std::is_arithmetic<T>::value &&
-                           !std::is_enum<T>::value &&
-                           sizeof(T) == sizeof(unsigned int) &&
-                           sizeof(unsigned int) == 4, bool> = true>
-RAJA_INLINE __device__ T hip_atomicLoad(T *acc)
-{
-  return RAJA::util::reinterp_A_as_B<unsigned int, T>(
-    hip_atomicLoad(reinterpret_cast<unsigned int*>(acc)));
-}
+/*!
+ * Alias for determining the integral type of the same size as the given type
+ */
+template <typename T>
+using hip_useReinterpretLoadStore_t = typename hip_useReinterpretLoadStore<T>::type;
 
 #endif
 
-#if defined(UINT64_MAX)
+
+/*!
+ * Atomic load
+ */
+#if defined(__has_builtin) && __has_builtin(__hip_atomic_load)
 
 template <typename T,
-          std::enable_if_t<!std::is_arithmetic<T>::value &&
-                           !std::is_enum<T>::value &&
-                           sizeof(T) == sizeof(uint64_t), bool> = true>
+          std::enable_if_t<hip_useIntrinsicLoadStore<T>::value, bool> = true>
 RAJA_INLINE __device__ T hip_atomicLoad(T *acc)
 {
-  return RAJA::util::reinterp_A_as_B<uint64_t, T>(
-    hip_atomicLoad(reinterpret_cast<uint64_t*>(acc)));
+  return __hip_atomic_load(acc, __ATOMIC_RELAXED, __HIP_MEMORY_SCOPE_AGENT);
 }
 
-#else
-
 template <typename T,
-          std::enable_if_t<!std::is_arithmetic<T>::value &&
-                           !std::is_enum<T>::value &&
-                           sizeof(T) == sizeof(unsigned long long int) &&
-                           sizeof(unsigned long long int) == 8, bool> = true>
+          std::enable_if_t<hip_useReinterpretLoadStore<T>::value, bool> = true>
 RAJA_INLINE __device__ T hip_atomicLoad(T *acc)
 {
-  return RAJA::util::reinterp_A_as_B<unsigned long long int, T>(
-    hip_atomicLoad(reinterpret_cast<unsigned long long int*>(acc)));
-}
+  using R = hip_useReinterpretLoadStore_t<T>;
 
-#endif
+  return RAJA::util::reinterp_A_as_B<R, T>(
+    hip_atomicLoad(reinterpret_cast<R*>(acc)));
+}
 
 #else
 
 template <typename T,
-          std::enable_if_t<std::is_same<T, int>::value ||
-                           std::is_same<T, unsigned int>::value ||
-                           std::is_same<T, unsigned long long int>::value, bool> = true>
+          std::enable_if_t<hip_useBuiltinCommon<T>::value, bool> = true>
 RAJA_INLINE __device__ T hip_atomicLoad(T *acc)
 {
-  return ::atomicOr(acc, static_cast<T>(0));
+  return hip_atomicOr(acc, static_cast<T>(0));
 }
 
 template <typename T,
-          std::enable_if_t<!std::is_same<T, int>::value &&
-                           !std::is_same<T, unsigned int>::value &&
-                           !std::is_same<T, unsigned long long int>::value &&
-                           sizeof(T) == sizeof(unsigned int), bool> = true>
+          std::enable_if_t<hip_useReinterpretCommon<T>::value, bool> = true>
```

Comment:
I wonder if it makes sense combining the reinterpret cases here since the implementation is the same and doing the ifdef just around the enable_if in the template.

Reply:
Done.